File: backend/app/core/upgrades/stat_modifiers/dash_cooldown_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class DashCooldown2(UpgradeImplementation):
    id = 'dash_cooldown_2'
    name = 'Swift Recovery'
    description = '-15% dash cooldown'
    rarity = 'epic'
    type = 'stat_modifier'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = -0.15
        appliers.modifiers.add_multiplier('player', 'dashCooldown', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/dash_cooldown_2.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const DashCooldown2Def: UpgradeDef = {
  id: "dash_cooldown_2",
  name: "Swift Recovery",
  description: "-15% dash cooldown",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 20,
  targetClass: UpgradeTargetID.Player,
  fieldInTargetClass: UpgradeStatID.DashCooldown,
  value: -0.15,
  isMultiplier: true,
  stackable: true,
  maxStacks: 3,
  dependentOn: [{ ids: ["dash_ability"] }],
}

export class DashCooldown2 extends Upgrade {}
